engine/runtime/src/main/java/org/enso/interpreter/service/ExecutionService.java

```diff
@@ -449,4 +454,53 @@ public String getExceptionMessage(PanicException panic) {
       context.getThreadManager().leave(p);
     }
   }
+
+  @SuppressWarnings("unchecked")
+  private static <E extends Exception> E raise(Class<E> type, Exception ex) throws E {
```

Comment:
Why do we need this method instead of putting `throw ex` directly at callsite?

Reply:
Okay, I see that for example `ArityException` or `UnsupportedTypeException` extend `InteropException` which apparently does not seem to inherit `RuntimeException`.

So I guess this is the reason for this magic?

But that seems totally unsafe - since we do an unchecked cast to `RuntimeException` of something that we know that it _doesn't_ extend `RuntimeException`.

If the cast is unchecked, I guess it will go through, but it will violate the typesystem rules. That sounds very worrying at first sight.

Can we get a comment (in the code, as part of this function's doc comment) explaining why we are doing it and why is it safe in this context? (If it is safe...)

Reply:
This is a standard way to propagate [checked exception](http://wiki.apidesign.org/wiki/Checked_exception) thru Java language. Java language has (too) strict rules about [checked exceptions](http://wiki.apidesign.org/wiki/Checked_exception), but there is nothing like that in the JVM (as demonstrated by Scala ignoring [checked exceptions](http://wiki.apidesign.org/wiki/Checked_exception) altogether). Sometimes it is useful to propagate [checked exception](http://wiki.apidesign.org/wiki/Checked_exception) even the Java method signatures are against it. This is the case. We have to tunnel the exception thru [CallTarget](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/CallTarget.html) interface now. Rather than wrapping the exception into runtime and unpacking it, I am trying to throw it directly. Btw. this is not the first place where this coding technique is used. There are other `raise` methods around the codebase already.